### DISCUSSION TITLE
How to implement Global Exception Handler

### DISCUSSION THREAD

**Discussion by michaelxswe** [fastapi/fastapi#11348] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.exception_handler(Exception)
async def common_exception_handler(request: Request, exc: Exception):
    return JSONResponse(
        status_code=500,
        content={"detail": "error"},
    )

@app.get("/")
async def test():
    print(1/0)
    return {"test":"test"}
```

### Description

I'm still getting error in asgi

```INFO:     127.0.0.1:49648 - "GET / HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/mx/Documents/Code/python/fast-api/.venv/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 411, in run_asgi
```

### Operating System

macOS

### FastAPI Version

latest

### Pydantic Version

latest

### Python Version

3.11.4

**Comment by PhysicallyActive** [comment]
A few pointers:
* Have you imported `Request` and `JSONResponse` (I had to manually add these to be able to run your code).
* Have you made sure that all of the dependencies are correctly linked?
* Can you provide some more details on the issue, e.g., the entire stack-trace?

This is a bit of an unconventional way to handle _all_ exceptions in an application as well, but if you want to do it; it could be better to use the [middleware functionality](https://fastapi.tiangolo.com/tutorial/middleware/):

```python
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware

app = FastAPI()


async def catch_exceptions_middleware(request: Request, call_next):
    try:
        return await call_next(request)
    except Exception as e:
        # you probably want some kind of logging here
        print(e)
        return JSONResponse(
            status_code=500,
            content={"detail": "error"},
        )


app.add_middleware(BaseHTTPMiddleware, dispatch=catch_exceptions_middleware)


@app.get("/")
async def test():
    _ = 1 / 0
    return {"test": "test"}
```

**Comment by YuriiMotov** [comment, marked as answer]
When you pass `Exception` class or `500` status code to `@app.exception_handler()` it installs this handler as a handler of `ServerErrorMiddleware`.
It's different from other exceptions and status codes that you can pass to `@app.exception_handler()`. Usual exceptions (inherited from `Exception`, but not 'Exception' itself) are handled by `ExceptionMiddleware` that just calls appropriate handlers (and doesn't raise this exception again).

Related part of FastAPI (Starlette) source code (`Starlette.build_middleware_stack`):
```
        for key, value in self.exception_handlers.items():
            if key in (500, Exception):
                error_handler = value
            else:
                exception_handlers[key] = value

        middleware = (
            [Middleware(ServerErrorMiddleware, handler=error_handler, debug=debug)]
            + self.user_middleware
            + [
                Middleware(
                    ExceptionMiddleware, handlers=exception_handlers, debug=debug
                )
            ]
        )
```

In the handler of `ServerErrorMiddleware` you can process the exception and return response, but `ServerErrorMiddleware` always raises that exception again.
Here is a related code from FastAPI (Starlette) sources that shows how `ServerErrorMiddleware` works:
```
class ServerErrorMiddleware:
    ...
    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        response_started = False

        async def _send(message: Message) -> None:
            nonlocal response_started, send

            if message["type"] == "http.response.start":
                response_started = True
            await send(message)

        try:
            await self.app(scope, receive, _send)
        except Exception as exc:
            request = Request(scope)
            if self.debug:
                # In debug mode, return traceback responses.
                response = self.debug_response(request, exc)
            elif self.handler is None:
                # Use our default 500 error handler.
                response = self.error_response(request, exc)
            else:
                # Use an installed 500 error handler.
                if is_async_callable(self.handler):
                    response = await self.handler(request, exc)
                else:
                    response = await run_in_threadpool(self.handler, request, exc)

            if not response_started:
                await response(scope, receive, send)

            # We always continue to raise the exception.
            # This allows servers to log the error, or allows test clients
            # to optionally raise the error within the test case.
            raise exc
```
